File: 1046-max-consecutive-ones-iii/1046-max-consecutive-ones-iii.cpp
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        int i = 0, j = 0, temp = 0, ans = 0;
        int N = nums.size();
        while (j < N) {
            if (nums[j] == 0)
                temp++;
            if (temp <= k) {
                ans = max(ans, j - i + 1);
                j++;
            } else {
                while (temp > k) {
                    if (nums[i] == 0)
                        temp--;
                    i++;
                }
                if (temp == k) {
                    ans = max(ans, j - i + 1);
                }
                j++;
            }
        }
        return ans;
    }
};

File: 1046-max-consecutive-ones-iii/1046-max-consecutive-ones-iii.java
class Solution {
    public int longestOnes(int[] nums, int k) {
        int i = 0;
        int j = 0;
        int count = 0;
        int ans = 0;
        while (j < nums.length) {
            if (nums[j] == 0) {
                count++;
            }
            if (count <= k) {
                ans = Math.max(ans, j - i + 1);
            } else if (count > k) {
                while (count > k) {
                    if (nums[i] == 0)
                        count--;
                    i++;
                }
                if (count == k) {
                    ans = Math.max(j - i + 1, ans);
                }
            }
            j++;
        }
        return ans;
    }
}